src/corona_hakab_model/generation/circles_generator.py

```diff
@@ -0,0 +1,131 @@
+        self.fill_social_circles()
+
+        # create whole population circles
+        self.create_whole_population_circles()
+
+        # export the population data
+        self.export()
+
+    def create_geographic_circles(self):
+        """
+        creates all geographic circles declared in generation consts.
+        each circle gets an object of GeographicalCircleDataHolder
+        :return:
+        """
+        for geo_circle in self.circles_consts.geographic_circles:
+            self.geographic_circles.append(GeographicCircle(geo_circle))
+
+    def allocate_agents(self):
+        """
+        splits the agents between the circles using each circle's agents share.
+        :return:
+        """
+        # making sure all agents shares sum up to one. if not, normalize them
+        share_factor = 1.0 / sum([geo_circle.data_holder.agents_share for geo_circle in self.geographic_circles])
+        # creating a dist for selecting a geographic circle for each agent
+        circle_selection = rv_discrete(values=(
+            range(len(self.geographic_circles)), [geo_circle.data_holder.agents_share * share_factor for geo_circle in self.geographic_circles]))
+        rolls = circle_selection.rvs(size=len(self.agents))
+        for agent, roll in zip(self.agents, rolls):
+            selected_geo_circle = self.geographic_circles[roll]
+            selected_geo_circle.add_agent(agent)
+            agent.geographic_circle = selected_geo_circle
+
+    def fill_social_circles(self):
+        """
+        fills self social circles by connection type from self geographic circles
+        :return:
+        """
+        for connection_type in ConnectionTypes:
+            for geo_circle in self.geographic_circles:
+                self.social_circles_by_connection_type[connection_type].extend(geo_circle.connection_type_to_social_circles[connection_type])
+
+    def create_whole_population_circles(self):
+        for connection_type in Whole_Population_types:
+            social_circle = SocialCircle(connection_type)
+            social_circle.add_many(self.agents)
+            self.social_circles_by_connection_type[connection_type].append(social_circle)
+
+    def export(self):
+        # fill population data with my data
+        self.population_data.agents = self.agents
+        self.population_data.social_circles_by_connection_type = self.social_circles_by_connection_type
+        self.population_data.geographic_circles = self.geographic_circles
+
+        # export population data using pickle
+        with open(self.EXPORT_OUTPUT_DIR + self.EXPORT_FILE_NAME, 'wb') as export_file:
+            pickle.dump(self.population_data, export_file)
+
+    def import_population_data(self, import_file_path=None):
+        if import_file_path is None:
+            import_file_path = self.EXPORT_OUTPUT_DIR + self.EXPORT_FILE_NAME
+
+        with open(import_file_path, 'rb') as import_file:
+            self.population_data = pickle.load(import_file)
```

Comment:
Notice this import won't work, because you don't use ```self.population_data``` at all - 
you should do something similar to the export function and re-assign all the relevant members.

Reply:
good point. fixed